### DISCUSSION TITLE
Sending a REST call from JavaScript Automation?

### DISCUSSION THREAD

**Discussion by Aculeo** [Budibase/budibase#10068] [Help]
I'm attempting to generate a PDF using an API from PDF.co, because Budibase cannot generate PDFs (yet). So for that I would need to call their API, get the response back, which contains a URL. I would like to get the file content, and write it back as an attachment into Budibase. Is this possible? Because I have noticed that `XMLHttpRequest` is not available in Automations. And using the available REST data connector would not allow me to write the file back to the DB.

To top it all of, and increasing my frustration further, I wanted to write an Azure Function, which can call PDF.co, and write back the PDF to Budibase. So far so good, except there does not appear to be a Public API for writing back attachments. So i'm at a wits end how I can make this work.

**Comment by CalvinArcurid** [comment]
Regarding your question, have you considered using a third-party service like Zapier or Integromat to connect PDF.co and Budibase? They have pre-built integrations that may be able to help you automate the process and avoid the need for XMLHttpRequest or an Azure Function.

If you do want to go the Azure Function route, have you considered using a file storage service like Azure Blob Storage or Amazon S3 to store the PDF and then linking to it in Budibase? That could be a workaround until a public API for writing back attachments becomes available.

I hope this helps! Let me know if you have any other questions or if there's anything else I can do to assist.

**Reply by Aculeo** [reply]
I've looked at Zapier, but their pricing model is very unattractive for such a small use case that I have - Integromat seems better though, I'll take a look at that, thanks! Still would mean I have to upload the generated PDF to some other place, as I can't upload it back to Budibase.

**Comment by ArchonMegalon** [comment]
This is still not implemented and not a valid solution. I for example would like to use my n8n-instance behind a cloudflare tunnel. I can set up a cloudflare service token, but i can only send ONE header in the n8n implementation of budibase (cloudflare requires clientid and clientsecret).

Well - no problem, just make the httprequest myself in js.

Only - the module is not there... :(

**Reply by ArchonMegalon** [reply]
at LEAST you guys could implement a http-call option as automation step.

**Reply by ArchonMegalon** [reply]
Oh - just found, that wget DOES work in bash scripting. all good! :)

**Reply by mike12345567** [reply]
Hi @ArchonMegalon - the best way to send requests from an automation is to use the external query step - https://docs.budibase.com/docs/external-data-connector, you can create a REST query in the data UI and then reference it here - this gives you complete control to query in whatever way you wish.

Adding async code to the JS step might be quite difficult due to the nature of the sandbox.
